### PR TITLE
Support array of arrays

### LINKED ISSUE #137: Array type support for array-of-arrays
**Description:**
Current implementation supports only int arrays at rust side and translation side.
We need to support for array-of-arrays and union-type.
    array-of-arrays, union-type

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**
#124

### PR DESCRIPTION
**Pull request by ramakotareddy**
## Purpose
Add support of two dimensional array with Int type.

Resolves #137  Array of Arrays 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: test/array/load_store_three_dim_array.bal [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// RUN: "%testRunScript" %s %nballerinacc "%java_path" "%skip_bir_gen" | filecheck %s
+
+public function print_string(string val) = external;
+
+public function printu32(int val) = external;
+
+public function main() {
+    int[][][] arr1 = [[], []];
+    arr1[0][1][1] = 1;
```
Comment (KavinduZoysa) [inline test/array/load_store_three_dim_array.bal:9]: How does three dim arrays are working in this PR?

File: compiler/BIRReader.cpp [hunk 2]
```diff
@@ -451,10 +451,12 @@ std::unique_ptr<ArrayInsn> ReadArrayInsn::readNonTerminatorInsn(std::shared_ptr<
 
     // TODO handle Array init values
     auto init_values_count = readerRef.readS4be();
+    std::vector<Operand> initValues;
```
Comment (ruvi-d) [inline compiler/BIRReader.cpp:454]: Delete the //TODO in line 452

Comment (ruvi-d) [inline compiler/BIRReader.cpp:454]: Since we know the size of the vector, use `reserve()` to set capacity

File: compiler/BIRReader.cpp [hunk 3]
```diff
@@ -451,10 +451,12 @@ std::unique_ptr<ArrayInsn> ReadArrayInsn::readNonTerminatorInsn(std::shared_ptr<
...
     for (size_t i = 0; i < init_values_count; i++) {
-        [[maybe_unused]] auto init_value = readerRef.readOperand();
+        auto initValue = readerRef.readOperand();
+        initValues.push_back(initValue);
```
Comment (ruvi-d) [inline compiler/BIRReader.cpp:457]: skip the temp named variable. `initValues.push_back(readerRef.readOperand());`

File: compiler/BIRReader.cpp [hunk 4]
```diff
@@ -451,10 +451,12 @@ std::unique_ptr<ArrayInsn> ReadArrayInsn::readNonTerminatorInsn(std::shared_ptr<
...
     }
-    return std::make_unique<ArrayInsn>(lhsOp, currentBB, sizeOperand);
+    return std::make_unique<ArrayInsn>(lhsOp, currentBB, sizeOperand, initValues);
```
Comment (ruvi-d) [inline compiler/BIRReader.cpp:459]: Add an early return to handle `init_values_count == 0` case. Better to use an `ArrayInsn` constructor without  initValues vector arg for that.

File: runtime/src/lib.rs [hunk 5]
```diff
@@ -282,6 +282,136 @@ pub extern "C" fn array_load_string(arr_ptr: *mut Vec<*mut String>, index: i32)
...
+pub extern "C" fn array_init_array_array(size: i32) -> *mut Vec<*mut Vec<*mut i32>> {
+    let size_t = if size > 0 { size } else { 8 };
+    let size_t = size_t as usize;
+    let foo: Box<Vec<*mut Vec<*mut i32>>> = Box::new(Vec::with_capacity(size_t));
+    let vec_pointer = Box::into_raw(foo);
+    return vec_pointer as *mut Vec<*mut Vec<*mut i32>>;
+}
+
+#[no_mangle]
+pub extern "C" fn array_init_array_int(size: i32) -> *mut Vec<*mut Vec<*mut i32>> {
```
Comment (ruvi-d) [inline runtime/src/lib.rs:295]: We need to first fix the issue where we are passing and store array member as references

File: runtime/src/lib.rs [hunk 6]
```diff
@@ -282,6 +282,136 @@ pub extern "C" fn array_load_string(arr_ptr: *mut Vec<*mut String>, index: i32)
     return return_val;
 }
 
+#[no_mangle]
+pub extern "C" fn array_init_array_array(size: i32) -> *mut Vec<*mut Vec<*mut i32>> {
+    let size_t = if size > 0 { size } else { 8 };
+    let size_t = size_t as usize;
+    let foo: Box<Vec<*mut Vec<*mut i32>>> = Box::new(Vec::with_capacity(size_t));
```
Comment (ruvi-d) [inline runtime/src/lib.rs:289]: use a meaningful variable name instead of `foo`. Repeat for other functions

File: runtime/src/lib.rs [hunk 7]
```diff
@@ -282,6 +282,136 @@ pub extern "C" fn array_load_string(arr_ptr: *mut Vec<*mut String>, index: i32)
     return return_val;
 }
 
+#[no_mangle]
+pub extern "C" fn array_init_array_array(size: i32) -> *mut Vec<*mut Vec<*mut i32>> {
```
Comment (ruvi-d) [inline runtime/src/lib.rs:286]: Why does both `array_init_array_int` and `array_init_array_array` both return `*mut Vec<*mut Vec<*mut i32>>` ?

File: runtime/src/lib.rs [hunk 8]
```diff
@@ -282,6 +282,136 @@ pub extern "C" fn array_load_string(arr_ptr: *mut Vec<*mut String>, index: i32)
...
+pub extern "C" fn array_init_array_bool(size: i32) -> *mut Vec<*mut Vec<*mut bool>> {
+    let size_t = if size > 0 { size } else { 8 };
+    let size_t = size_t as usize;
+    let foo: Box<Vec<*mut Vec<*mut bool>>> = Box::new(Vec::with_capacity(size_t));
+    let vec_pointer = Box::into_raw(foo);
+    return vec_pointer as *mut Vec<*mut Vec<*mut bool>>;
+}
+
+#[no_mangle]
+pub extern "C" fn array_init_array_string(size: i32) -> *mut Vec<*mut Vec<*mut String>> {
```
Comment (ruvi-d) [inline runtime/src/lib.rs:322]: String needs to use BString

File: runtime/src/lib.rs [hunk 9]
```diff
@@ -282,6 +282,136 @@ pub extern "C" fn array_load_string(arr_ptr: *mut Vec<*mut String>, index: i32)
...
+    let foo: Box<Vec<*mut Vec<*mut String>>> = Box::new(Vec::with_capacity(size_t));
+    let vec_pointer = Box::into_raw(foo);
+    return vec_pointer as *mut Vec<*mut Vec<*mut String>>;
+}
+
+#[no_mangle]
+pub extern "C" fn array_load_array_int(arr_ptr: *mut Vec<*mut Vec<*mut i32>>, index: i32) -> *mut Vec<*mut i32> {
+    let mut arr = unsafe { Box::from_raw(arr_ptr) };
+    let index_n = index as usize;
+    let len = index_n + 1;
```
Comment (ruvi-d) [inline runtime/src/lib.rs:334]: assert that index is not negative

File: runtime/src/lib.rs [hunk 10]
```diff
@@ -282,6 +282,136 @@ pub extern "C" fn array_load_string(arr_ptr: *mut Vec<*mut String>, index: i32)
...
+    if return_val.is_null() {
+        arr[index_n] = Box::into_raw(Box::new(Vec::with_capacity(len)));
+    }
+    return_val = arr[index_n];
+    mem::forget(arr);
+    return return_val;
+}
+
+#[no_mangle]
+pub extern "C" fn array_load_array_array(arr_ptr: *mut Vec<*mut Vec<*mut i32>>, index: i32) -> *mut Vec<*mut i32> {
```
Comment (ruvi-d) [inline runtime/src/lib.rs:399]: Same questions as with the corresponding init function, why is this same as `array_load_array_int` ?

File: runtime/src/lib.rs [hunk 11]
```diff
@@ -282,6 +282,136 @@ pub extern "C" fn array_load_string(arr_ptr: *mut Vec<*mut String>, index: i32)
...
+    }
+    let mut return_val = arr[index_n];
+    if return_val.is_null() {
+        arr[index_n] = Box::into_raw(Box::new(Vec::with_capacity(len)));
+    }
+    return_val = arr[index_n];
+    mem::forget(arr);
+    return return_val;
+}
+
```
Comment (ruvi-d) [inline runtime/src/lib.rs:414]: Missing deinit functions

File: test/array/twod_int_array_with_any_typecast.bal [hunk 12]
```diff
@@ -0,0 +1,17 @@
+// RUN: "%testRunScript" %s %nballerinacc "%java_path" | filecheck %s
```
Comment (ruvi-d) [inline test/array/twod_int_array_with_any_typecast.bal:1]: please check test file name. format should be `<action>_<type>_<other>.bal`

File: test/array/twod_int_array_with_any_typecast.bal [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+public function print_string(string val) = external;
+
+public function printu32(int val) = external;
+
+public function main() {
+    int[5][3] arr4 = [[], []];
+    arr4[0][0] = 10;
+    arr4[1][2] = 20;
+    any i1 = arr4[0][0];
```
Comment (ruvi-d) [inline test/array/twod_int_array_with_any_typecast.bal:11]: for casting to any, we should be casting the entire multi dimensional array. as in `any val = arr4;`

File: runtime/src/lib.rs (hunk 11 above)
Comment (ruvi-d) [inline runtime/src/lib.rs:414]: Missing tests for array init values that are not arrays. e.g. int, float, bool and string types.

File: test/array/string_twodim_array_with_global_var.bal [hunk 14]
```diff
@@ -0,0 +1,13 @@
+// RUN: "%testRunScript" %s %nballerinacc "%java_path" | filecheck %s
```
Comment (ruvi-d) [inline test/array/string_twodim_array_with_global_var.bal:1]: please check test file name. format should be `<action>_<type>_<other>.bal`

File: test/array/string_twodim_array_with_any_typecast.bal [hunk 15]
```diff
@@ -0,0 +1,18 @@
+// RUN: "%testRunScript" %s %nballerinacc "%java_path" | filecheck %s
```
Comment (ruvi-d) [inline test/array/string_twodim_array_with_any_typecast.bal:1]: please check test file name. format should be `<action>_<type>_<other>.bal`

File: test/array/load_store_two_dim_array_boolean.bal [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+public function print_boolean(boolean val) = external;
+
+public function main() {
+    boolean[][] arr1 = [[], []];
+    arr1[0][1] = true;
+    arr1[1][1] = false;
+    arr1[3][3] = true;
+    int index1 = 3;
+    boolean b = arr1[index1][index1];
+    if (b) {
```
Comment (ruvi-d) [inline test/array/load_store_two_dim_array_boolean.bal:12]: need to check at least 2 member values from array

File: runtime/src/lib.rs (hunk 11 above)
Comment (ruvi-d) [inline runtime/src/lib.rs:414]: Missing test to cover function's with multi dimensional array return types

**Review by ruvi-d — CHANGES_REQUESTED** [review]

File: runtime/src/lib.rs (hunk 11 above)
Comment (KavinduZoysa) [reply runtime/src/lib.rs:414]: Shall we add a test case something like this, 
```
function foo() returns string[][] {
    string[][] s = [[], []];
    s[0][0] = "Test";
    return s;
}

public function main() {
    string[][] s = foo();
    string x = s[0][0];
}
```

Comment (ramakotareddy) [reply runtime/src/lib.rs:414]: @ruvi-d, Please give any simple testcase.

Comment (ruvi-d) [reply runtime/src/lib.rs:414]: `int arr1[] = [1,2,3,4];` for float, bool and string

File: runtime/src/lib.rs (hunk 5 above)
Comment (ramakotareddy) [reply runtime/src/lib.rs:295]: Fixed in #242 PR.

File: runtime/src/lib.rs (hunk 8 above)
Comment (ramakotareddy) [reply runtime/src/lib.rs:322]: Fixed in #242 PR.

**Pull request closed without merging**